src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -336,64 +367,98 @@ namespace cryptonote
     LOG_INFO_CC(context, "New connection posing as pruning seed " << epee::string_tools::to_string_hex(context.m_pruning_seed) << ", seed address " << &context.m_pruning_seed);
 #endif
 
+    // No chain synchronization over hidden networks (tor, i2p, etc.)
+    if(context.m_remote_address.get_zone() != epee::net_utils::zone::public_)
+    {
+      context.m_state = cryptonote_connection_context::state_normal;
+      return true;
+    }
+
+    auto curr_height = m_core.get_current_blockchain_height();
+
+    context.m_need_blink_sync = false;
     // Check for any blink txes being advertised that we don't know about
     {
-      // Do this in two passes: first pass just validates and quits on failure
-      uint64_t last_height = 0;
-      for (auto &b : hshd.blink_txs) {
-        if (b.height > last_height) {
-          last_height = b.height;
-        } else {
+      if (hshd.blink_blocks.size() != hshd.blink_hash.size())
+      {
+        MWARNING(context << " peer sent illegal mismatched blink heights/hashes; disconnecting");
+        return false;
+      }
+      else if (hshd.blink_blocks.size() > 1000)
+      {
+        MWARNING(context << " peer sent too many post-checkpoint blink blocks; disconnecting");
+        return false;
+      }
+
+      // Peer sends us HEIGHT -> HASH pairs, where the HASH is the xor'ed tx hashes of all blink
+      // txes mined at the given HEIGHT.  If the HASH is different than our hash for the same height
+      // *and* different than the last height the peer sent then we will request the blink txes for
+      // that height.
+
+      const uint64_t immutable_height = m_core.get_blockchain_storage().get_immutable_height();
+      context.m_blink_state.erase(context.m_blink_state.begin(), context.m_blink_state.lower_bound(immutable_height));
+      auto our_blink_hashes = m_core.get_pool().get_blink_checksums();
+      uint64_t last_height;
+      for (size_t i = 0; i < hshd.blink_blocks.size(); i++) {
+        auto &height = hshd.blink_blocks[i];
+        if (i == 0 || height > last_height)
+          last_height = height;
+        else {
           MWARNING(context << " peer sent blink tx heights out of order, which is not valid; disconnecting");
           return false;
         }
 
-        if (b.hashes.empty()) {
-          MWARNING(context << " peer sent invalid blink tx hash data, disconnecting");
-          return false;
+        if (height > 0 && (height < immutable_height || height >= curr_height))
+        {
+          // We're either past the immutable height (in which case we don't care about the blink
+          // signatures), or we don't know about the advertised block yet (we'll get the blink info
+          // when we get the block).  Skip it but don't disconnect because this isn't invalid.
+          continue;
```

Comment:
The reason to continue here in that case is to detect and log an error if a peer sends things out of order, so that if it sends something like [0, 456, 455] we can warn (to notice a problem) and disconnect the peer (for doing something wrong).  If everything is valid the check is cheap - we will only end up doing a few integer comparisons for the rest of the loop.

Reply:
okay, makes sense.